File: platforms/ios/Dezel/Dezel/Modules/Web/LocalStorage.swift
/**
 * @class LocalStorage
 * @since 0.1.0
 * @hidden
 */
open class LocalStorage: JavaScriptObject {

	//--------------------------------------------------------------------------
	// MARK: JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_removeItem
	 * @since 0.1.0
	 * @hidden
	 */
	@objc open func jsFunction_removeItem(callback: JavaScriptFunctionCallback) {

		let count = callback.arguments
		if (count < 1) {
			return
		}

		let key = callback.argument(0).string

		UserDefaults.standard.removeObject(forKey: kPrefix + key)
	}

	/**
	 * @method jsFunction_setItem
	 * @since 0.1.0
	 * @hidden
	 */
	@objc open func jsFunction_setItem(callback: JavaScriptFunctionCallback) {

		if (callback.arguments < 2) {
			return
		}

		let key = callback.argument(0).string
		let val = callback.argument(1).string

		UserDefaults.standard.set(val, forKey: kPrefix + key)
	}

	/**
	 * @method jsFunction_getItem
	 * @since 0.1.0
	 * @hidden
	 */
	@objc open func jsFunction_getItem(callback: JavaScriptFunctionCallback) {

		let count = callback.arguments
		if (count < 1) {
			return
		}

		let key = callback.argument(0).string

		if let val = UserDefaults.standard.string(forKey: kPrefix + key) {
			callback.returns(string: val)
		}
	}
}

private let kPrefix = "dezel.localStorage."

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/modules/web/LocalStorage.kt
package ca.logaritm.dezel.modules.web

import android.content.SharedPreferences
import ca.logaritm.dezel.application.application
import ca.logaritm.dezel.core.JavaScriptClass
import ca.logaritm.dezel.core.JavaScriptContext
import ca.logaritm.dezel.core.JavaScriptFunctionCallback

/**
 * @class LocalStorage
 * @since 0.1.0
 * @hidden
 */
open class LocalStorage(context: JavaScriptContext) : JavaScriptClass(context) {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property preferences
	 * @since 0.1.0
	 * @hidden
	 */
	private val preferences: SharedPreferences = this.context.application.getSharedPreferences("dezel.localStorage", android.content.Context.MODE_PRIVATE)

	//--------------------------------------------------------------------------
	// JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_removeItem
	 * @since 0.1.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_removeItem(callback: JavaScriptFunctionCallback) {

		val count = callback.arguments
		if (count < 1) {
			return
		}

		val editor = this.preferences.edit()
		if (editor == null) {
			return
		}

		val storageKey = callback.argument(0).string

		editor.remove(storageKey)
		editor.apply()
	}

	/**
	 * @method jsFunction_setItem
	 * @since 0.1.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_setItem(callback: JavaScriptFunctionCallback) {

		val count = callback.arguments
		if (count < 2) {
			return
		}

		val editor = this.preferences.edit()
		if (editor == null) {
			return
		}

		val storageKey = callback.argument(0).string
		val storageVal = callback.argument(1).string

		editor.putString(storageKey, storageVal)
		editor.apply()
	}

	/**
	 * @method jsFunction_getItem
	 * @since 0.1.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_getItem(callback: JavaScriptFunctionCallback) {

		val count = callback.arguments
		if (count < 1) {
			return
		}

		val key = callback.argument(0).string

		val value = this.preferences.getString(key, null)
		if (value != null) {
			callback.returns(value)
		}
	}
}
